### PR TITLE
[ENH] Updated EncoderRegressor

### PR DESCRIPTION
**Pull request by AnonymousCodes911**
#### Reference Issues/PRs
fixes: #1218 

#### What does this implement/fix? Explain your changes.
Adapts `EncoderClassifier` to `EncoderRegressor`

#### Any other comments?
Closes: #1204 

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).

##### For developers with write access
- [ ] Optionally, I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/regression/deep_learning/_encoder.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        fc_units=256,
+        callbacks=None,
+        file_path="./",
+        save_best_model=False,
+        save_last_model=False,
+        best_file_name="best_model",
+        last_file_name="last_model",
+        verbose=False,
+        loss="mean_squared_error",
+        metrics=("mean_squared_error",),
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_encoder.py:102]: metrics should be None by default and then if the user does not change it, we set a value for it when its used inside `_fit`

File: aeon/regression/deep_learning/_encoder.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        return model
+
+    def _fit(self, X, y):
+        """Fit the classifier on the training set (X, y).
+
+        Parameters
+        ----------
+        X : np.ndarray of shape = (n_instances (n), n_channels (d), series_length (m))
+            The training input samples.
+        y : np.ndarray of shape n
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_encoder.py:202]: its not class labels its target values

File: aeon/regression/deep_learning/_encoder.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        """
+        import tensorflow as tf
+
+        # Transpose X to conform to Keras input style
+        X = X.transpose(0, 2, 1)
+        check_random_state(self.random_state)
+
+        self.input_shape = X.shape[1:]
+        self.training_model_ = self.build_model(self.input_shape, 1)
+        # changed n_classes to 1
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_encoder.py:217]: no need just remove the n_classes parameter

File: aeon/regression/deep_learning/_encoder.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    }
+
+    def __init__(
+        self,
+        n_epochs=100,
+        batch_size=12,
+        kernel_size=None,
+        n_filters=None,
+        dropout_proba=0.2,
+        activation="linear",
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_encoder.py:89]: activation is the activation of the hidden layers, in the case of regression we add a new parameter output_activation and we set it to linear by default

File: aeon/regression/deep_learning/_encoder.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        )
+
+        return model
+
+    def _fit(self, X, y):
+        """Fit the classifier on the training set (X, y).
+
+        Parameters
+        ----------
+        X : np.ndarray of shape = (n_instances (n), n_channels (d), series_length (m))
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_encoder.py:200]: can you change this to (n_instances, n_channels, n_timepoints)

File: aeon/regression/deep_learning/_encoder.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        self._network = EncoderNetwork(
+            kernel_size=self.kernel_size,
+            max_pool_size=self.max_pool_size,
+            n_filters=self.n_filters,
+            fc_units=self.fc_units,
+            strides=self.strides,
+            padding=self.padding,
+            dropout_proba=self.dropout_proba,
+            activation=self.activation,
+            random_state=self.random_state,
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_encoder.py:145]: can you merge with main ? we just got in the removal of random_state in networks module

File: aeon/regression/deep_learning/_encoder.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+from aeon.regression.deep_learning.base import BaseDeepRegressor
+
+
+class EncoderRegressor(BaseDeepRegressor):
+    """
+    Establishing the network structure for an Encoder.
+
+    Adapted from the implementation used in classification.deeplearning
+
+    Parameters
```
Comment (hadifawaz1999) [inline aeon/regression/deep_learning/_encoder.py:23]: a lot of parameters are missing from the docstring

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
Thanks for this. All of the mentioned comments can be solved by comparing FCNClassifier and FCNRegressor as a baseline to know how to create EncodeRegressor

File: aeon/regression/deep_learning/_encoder.py (hunk 7 above)
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py:23]: Added the missing parameters

File: aeon/regression/deep_learning/_encoder.py (hunk 2 above)
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py:209]: changed `class labels` to `Target Value`

File: aeon/regression/deep_learning/_encoder.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+            n_filters=self.n_filters,
+            fc_units=self.fc_units,
+            strides=self.strides,
+            padding=self.padding,
+            dropout_proba=self.dropout_proba,
+            activation=self.activation,
+            random_state=self.random_state,
+        )
+
+    def build_model(self, input_shape, n_classes, **kwargs):
```
[the remark this answers is not in the document]
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py]: removed the parameter `n_classes`

File: aeon/regression/deep_learning/_encoder.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+        In aeon, time series are stored in numpy arrays of shape (d, m), where d
+        is the number of dimensions, m is the series length. Keras/tensorflow assume
+        data is in shape (m, d). This method also assumes (m, d). Transpose should
+        happen in fit.
+
+        Parameters
+        ----------
+        input_shape : tuple
+        The shape of the data fed into the input layer, should be (m, d).
+        n_classes : int
```
[the remark this answers is not in the document]
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py]: removed the definition of `n_classes` from the documentation.

File: aeon/regression/deep_learning/_encoder.py (hunk 1 above)
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py]: changed metrics to `None`

File: aeon/regression/deep_learning/_encoder.py (hunk 3 above)
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py]: removed the `n_classes` parameter

File: aeon/regression/deep_learning/_encoder.py (hunk 4 above)
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py]: added the new parameter `output_activation`

File: aeon/regression/deep_learning/_encoder.py (hunk 5 above)
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py]: done as directed.

File: aeon/regression/deep_learning/_encoder.py (hunk 6 above)
Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py]: removed `random_state`

**Review by hadifawaz1999 — CHANGES_REQUESTED** [review]
Thanks it all looks good now, last thing to modify is:

1- merge branch main
2- do the changes on model saving as we did yesterday in #1292  from hdf5 to keras format

File: aeon/regression/deep_learning/_encoder.py
Comment (AnonymousCodes911) [inline aeon/regression/deep_learning/_encoder.py:1]: @hadifawaz1999 , did the needful ! thank you!

Comment (hadifawaz1999) [reply aeon/regression/deep_learning/_encoder.py:1]: Could you merge main branch with this one ? thanks

Comment (AnonymousCodes911) [reply aeon/regression/deep_learning/_encoder.py:1]: its already up to date!

**Review by hadifawaz1999 — APPROVED** [review]
lgtm !
